### DISCUSSION TITLE
Solving the greatest-n-per-group problem

### DISCUSSION THREAD

**Discussion by seijikun** [diesel-rs/diesel#4127] [Q&A]
This is in relation to this [Reddit post](https://www.reddit.com/r/rust/comments/1e8ld5d/my_take_on_databases_with_rust_seaorm_vs_diesel/)
One of the problems I was attempting to solve as a Diesel newcomer is the famous `greatest-n-per-group` problem.

**Backend:** `Sqlite`

Following example relations:
```mermaid
erDiagram
    Customer ||--o{ Contract : "customer_id"
    Contract {
        id UUID
        customer_id UUID
    }
    CustomerAddress {
        id UUID
        customer_id UUID
        valid_from DATE
    }
    Customer {
        id UUID
    }

    CustomerAddress ||--o{ Customer : "id"
```
```sql
CREATE TABLE "CustomerAddress" (
    id UUID NOT NULL;
    customer_id UUID REFERENCES Customer("id");
    valid_from DATE NOT NULL;
);
CREATE TABLE "Customer" (
    id UUID NOT NULL;
);
CREATE TABLE "Contract" (
    id UUID NOT NULL;
    customer_id UUID NOT NULL REFERENCES Customer("id");
);
```

I have a list of contracts. Every contract is between me and a customer. The customer's address (where the invoice should be sent, e.g.) changes over time. Each `CustomerAddress` has a date specifying the starting point (as a `date`) from which on it is valid.

Getting the currently valid address for a given customer is easy using SQL:
```sql
SELECT CustomerAddress.* FROM CustomerAddress
WHERE valid_from <= current_date AND customer_id = ?
GROUP BY customer_id
-- greatest-n-per-group, only yields 1 result if UNIQUE(customer_id, valid_from)
HAVING valid_from = max(valid_from)
```
Now, representing my version of the currently valid address in diesel is not possible, since PostgreSQL doesn't like that. The alternative version is this:
```sql
SELECT caddr1.* FROM customer_addresses caddr1
  JOIN (
      SELECT customer_id, MAX(valid_from) AS valid_from
      FROM customer_addresses
      WHERE valid_from <= current_date
      GROUP BY customer_id
  ) caddr2 ON caddr1.customer_id = caddr2.customer_id AND caddr1.valid_from = caddr2.valid_from;
```
Which is already rather unwieldy in comparison.
My attempt of representing this second version with diesel failed at the `MAX(valid_from) AS valid_from`, because I was unable to find a way to select an expression with an alias name.
Unfortunately, my attempts of implementing this with Diesel are no longer with us at this point. I can just provide some fragments:
The variant (I think) that was closest to working went something like this:

```rust
let ca2 = alias!(customer_addresses as ca2);
let max_valid_from = dsl::sql::<sql_types::Date>("MAX(valid_from) as valid_from");
let ca2_customer_id = ca2.field(customer_addresses::customer_id);
let ca2_valid_from = ca2.field(customer_addresses::valid_from);

let ca2_subquery = ca2
	.select((ca2_customer_id, max_valid_from))
	.filter(ca2_valid_from.le(date))
	.group_by(ca2_customer_id)
	.load::<(EntityId, time::Date)>(con);

customer_addresses::table
	.select(CustomerAddress::as_select())
	.inner_join(
		ca2_subquery
			.on(customer_addresses::customer_id.eq(ca2_customer_id))
			.on(customer_addresses::valid_from.eq(ca2_valid_from))
	);
```

This was the problem I spent hours to try to solve, before finally giving up. Shame on me, if there is an easy solution.

---
**Addendum:**
The actual use-case, however, is to get a list of all contracts, joined with all the attached information (`Customer`, and currently valid `CustomerAddress`):

In raw SQL, that's:
```sql
SELECT * FROM "Contract"
-- (some more INNER JOINS in actual use-case)
INNER JOIN "Customer" ON "Customer"."id" = "Contract"."customer_id"
INNER JOIN (
   ...
) addrjoin ON addrjoin.customer_id = Contract.customer_id
```

**Comment by weiznich** [comment]
Thanks for opening this discussion. I spend a bit time on looking into this today. First of all I want to note that this is in my opinion already a bit outside the simple query part, which makes it more likely that you just hit something that is not possible to express with the built-in DSL yet.

Now to the actual question: 

> ```sql 
> SELECT CustomerAddress.* FROM CustomerAddress
> WHERE valid_from <= current_date AND customer_id = ?
> GROUP BY customer_id
> -- greatest-n-per-group, only yields 1 result if UNIQUE(customer_id, valid_from)
> HAVING valid_from = max(valid_from)
> ```

The important part here is your comment in your query: It says if that constraint is true this query does not return non-deterministic results. Diesel does not know about that constraint and therefore rejects such constructs.

> Now, representing my version of the currently valid address in diesel is not possible, since PostgreSQL doesn't like that. The alternative version is this:
> ```sql
> SELECT caddr1.* FROM customer_addresses caddr1
>  JOIN (
>      SELECT customer_id, MAX(valid_from) AS valid_from
>      FROM customer_addresses
>      WHERE valid_from <= current_date
>      GROUP BY customer_id
>  ) caddr2 ON caddr1.customer_id = caddr2.customer_id AND caddr1.valid_from = caddr2.valid_from;
> ```

This uses a join to a subquery, which is something that's not supported by the built-in DSL yet. You can likely make that work via a custom DSL extension (see the extending diesel guide for details) or by just using `diesel::sql_query` (It's there to be used in such cases).

Now there is another solution to this problem: Use a subquery:

```sql
SELECT `customer_address`.`id`, `customer_address`.`customer_id`, `customer_address`.`valid_from` 
FROM `customer_address` 
WHERE (`customer_address`.`valid_from` = (
    SELECT max(`ca`.`valid_from`) 
    FROM `customer_address` AS `ca` 
    WHERE ((`ca`.`customer_id` = `customer_address`.`customer_id`) AND (`ca`.`valid_from` < CURRENT_DATE)) LIMIT ?)
) 
```

This can be expressed by diesels DSL in the following way:

```rust
    let ca = diesel::alias!(customer_address as ca);

    let max_sub_query = ca
        .select(diesel::dsl::max(ca.field(customer_address::valid_from)))
        .filter(
            ca.field(customer_address::customer_id)
                // the crucial part here is to reference the relevant
                // column from the outer query
                .eq(customer_address::customer_id),
        )
        .filter(
            ca.field(customer_address::valid_from)
                .lt(diesel::dsl::today),
        );

    // add other filters here as required 
    let result = customer_address::table.filter(
        customer_address::valid_from
            .nullable()
            .eq(max_sub_query.single_value()),
    ).load(&conn);

```

> The actual use-case, however, is to get a list of all contracts, joined with all the attached information (Customer, and currently valid CustomerAddress):

That should be possible, but might need another level of subqueries. I think the important "trick" to remember here is that you can turn most joins into a subquery, which is something that's expressible in diesel.

**Comment by seijikun** [comment]
Thanks for the response!

> First of all I want to note that this is in my opinion already a bit outside the simple query part

I don't know whether my use-case would be called simple. I have to admit, I was a bit surprised that I couldn't find another discussion asking for named expressions (`MAX(test) as newname`) in the select clause, or subqueries in joins, since I personally would regard both as important basic buildings blocks that I use a lot (I ran into this problem within hours after starting). I just hoped I'm too dumb to find them. So maybe I'm the problem.

>This uses a join to a subquery, which is something that's not supported by the built-in DSL yet. You can likely make that work via a custom DSL extension (see the extending diesel guide for details) or by just using diesel::sql_query (It's there to be used in such cases).

So, as far as I understood it - `sql_query` is basically like `sqlx`: You're on your own, and there is none of the query builder functionality left (like adding optional WHERE and ORDER BY clauses), right? If so, that's not really an option for me, because I was aiming for said query to be composable.
That leaves me with:

---
**Extending the `dsl` to support subqueries in `inner_join()`:**

I've had quite the struggle with understanding the generic bounds for the entire join story.
For simplicity, my idea to tackle this was to allow using any `SqlQuery` constructed by `diesel::sql_query("...")` in `inner_join()` with a custom `ON` clause and an alias name (`AS`).
And, in case diesel needs some type guarantees or type information, to just provide them by hand for now.

I started by creating a struct
```rust
struct SubQueryAs {
    query: SqlQuery,
    alias: String
}
```
I found that `JoinWithImplicitOnClause` accepts anything where `Lhs: JoinTo<Rhs>`.
That means, that `Lhs` (the rest of the query) needs an implementation of `JoinTo` for my subquery.
... which I can't provide, due to the orphan rule.

Diesel's internal `OnClauseWrapper` is public via `helper_types::On`, so I went that way instead, since for that type, implementations of `JoinTo<OnClauseWrapper<Rhs, On>> exist.
A simple:
```rust
impl SubQueryAs {
    pub fn on<On>(self, on: On) -> helper_types::On<SubQueryAs, On> {
        helper_types::On::new(self, on)
    }
}
```
allowed me to then do this:
```rust
let subquery = SubQueryAs {
	query: diesel::sql_query("SELECT * FROM customer_addresses"),
	alias: "addrjoin".into()
}.on(customer_addresses::customer_id.eq(rental_agreements::customer_id));
```
Using it inside `inner_join()` then requires me to implement `QuerySource` for `SubQueryAs`, which I quickly hacked together.
```rust
impl QuerySource for SubQueryAs {
    type FromClause = <customer_addresses::table as QuerySource>::FromClause;
    type DefaultSelection = <customer_addresses::table as QuerySource>::DefaultSelection;

    fn from_clause(&self) -> Self::FromClause {
        customer_addresses::table.from_clause()
    }

    fn default_selection(&self) -> Self::DefaultSelection {
        customer_addresses::all_columns
    }
}
// In hopes of not having to provide further type info, I went with Untyped
impl diesel::Expression for SubQueryAs {
    type SqlType = diesel::expression::expression_types::Untyped;
}
```
then I attempted to implement `SelectableExpression` for the columns I was interested in (for now):
```rust
impl SelectableExpression<SubQueryAs> for customer_addresses::columns::id {}
```
but I got stuck with errors like:
```
type mismatch resolving `<SubQueryAs as AppearsInFromClause<table>>::Count == Once`
associated types for the current `impl` cannot be restricted in `where` clauses

schema.rs(5, 9): required for `schema::customer_addresses::columns::id` to implement `AppearsOnTable<SubQueryAs>`
mod.rs(325, 45): required by a bound in `diesel::SelectableExpression`
the trait bound `SubQueryAs: TableNotEqual<schema::customer_addresses::table>` is not satisfied
double check that `schema::customer_addresses::table` and `SubQueryAs` appear in the same `allow_tables_to_appear_in_same_query!`
```
at which point I started to question whether I was going in the right direction.
Do you have a hint for me on how to proceed?

**Reply by weiznich** [reply]
First of all I want to write that this is not the way that I suggested as preferred solution. It would be much easier to go the subquery way as demonstrated in the post above.

Additionally I can add that the approach chosen by you does not really follow that what's demonstrated in the "Extending Diesel" guide. That guide shows how you wrap a subquery to write down a specific query, not a fully generic thing. That's not by accident, as any additional generically makes things much harder. I cannot say you shouldn't do that, but please don't expect that it makes things easier nor that I spend time in helping you figuring out the in my opinion wrong approach to address this problem (again: look at the subquery based solution, that works out of the box). For this specific case this would mean that you would wrap your inner query and use the outer `QueryFragment` impl to write down the specific query you have in mind.
